Add `gen plugin` command to generate boilerplate for plugins

johnSchnake opened the pull request:
When creating a plugin there are two main steps:
 - generating the plugin manifest
 - putting that manifest into the `sonobuoy gen` YAML to run it

This addresses the first point. Using this command we can generate
the boilerplate YAML for most standard plugins. It allows
customizing the plugin type, name, image, command, and env vars.

Fixes #665

Signed-off-by: John Schnake <jschnake@vmware.com>

**Release note**:
```
Added `gen plugin` command to generate plugin manifest for custom plugins.
```

johnSchnake commented on cmd/sonobuoy/app/plugindriver.go:
```diff
@@ -0,0 +1,44 @@
+		return fmt.Errorf("unknown plugin driver %q", str)
+	}
+	fmt.Println("schnake driver map", driverMap)
+	fmt.Println("schnake str", str, "driver", driver, ok)
```
Remove debug lines...

johnSchnake commented on pkg/plugin/manifest/manifest.go:
```diff
@@ -44,7 +44,7 @@ func (s *SonobuoyConfig) DeepCopy() *SonobuoyConfig {
 	SonobuoyConfig SonobuoyConfig `json:"sonobuoy-config"`
 	Spec           Container      `json:"spec"`
-	ExtraVolumes   []Volume       `json:"extra-volumes"`
+	ExtraVolumes   []Volume       `json:"extra-volumes,omitempty"`
```
This was because otherwise, we got a slightly ugly (and unnecessary) `extra-volumes: []` which we didn't normally see in our manually written YAML.

johnSchnake commented on cmd/sonobuoy/app/gen_plugin_def.go:
```diff
@@ -0,0 +1,144 @@
+}
+
+// toYAML serializes the manifest to YAML.
+func toYAML(m manifest.Manifest) ([]byte, error) {
```
No point to having this method, its completely trivial (only found that out after creating it).

johnSchnake commented on cmd/sonobuoy/app/gen_plugin_def.go:
```diff
@@ -0,0 +1,144 @@
+// on top of.
+func defaultManifest() manifest.Manifest {
+	m := manifest.Manifest{}
+	m.Spec.Name = "plugin"
```
Use consts/vars for these magic strings/values.

johnSchnake commented:
Good to see the test stats. Looking at the lines covered/not it is not surprising: the CLI based code (setting up commands, handling input, etc) is not tested well while the actual logic is covered fine.

Some refactoring will need to be done to try and easily test more of the codebase (right now its too awkward to test all the CLI stuff because we have exits throughout the code (among a few other things that could be fixed up).

chuckha commented on cmd/sonobuoy/app/gen_plugin_def.go:
```diff
@@ -0,0 +1,146 @@
+
+// genPluginDef returns the YAML for the plugin which Sonobuoy would expect as
+// a configMap in order to run/gen a typical run.
+func genPluginDef(cfg *GenPluginDefConfig) (string, error) {
```
I kind of expected this to have a return type of `([]byte, error)`, but I can see why this works as `(string, error)`

johnSchnake commented:
Changes since approval:
 - changed signature to use ([]byte,error) which is more idiomatic (and fixed references)
 - changed capitalization error I found with plugin driver (`Daemonset` was used instead of `DaemonSet`)
 - because of how annoyed I was at the issue above, I changed the plugin loader to use `ToLower` before comparing the plugindriver. We do not intend for there to be a `Job` and a `job` meaning different things. I could easily see someone *cough* me *cough* making that mistake again
